core/state/snapshot/snap_iterator.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>
+
+package snapshot
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb"
+)
+
+// snapIter is a wrapper of underlying database iterator. It extends
+// the basic iteration interface by adding Discard which can cache
+// the element internally where the iterator is currently located.
+type snapIter struct {
+	it        ethdb.Iterator
+	key       []byte
+	val       []byte
+	checkLast bool
+}
+
+// newSnapIter initializes the snapshot iterator.
+func newSnapIter(it ethdb.Iterator) *snapIter {
+	return &snapIter{it: it}
+}
+
+// Discard stores the element internally where the iterator is currently located.
+func (it *snapIter) Discard() {
```

Comment:
Itt took me a very long time to figure out what this method is for. I really really think we need a better name for it. By discard I would assume that an item is skipped, but this method does the exact opposite, it "undoes" on step. Alas, it can't undo multiple, so `undo` is a bad name too. Perhaps `Hold`? That would somehwat resemble the operation of "keeping hold of" the current item and maybe serving it up the next time too?

Reply:
Yup, sounds a good name to me.